javaparser-testing/src/test/java/com/github/javaparser/utils/VisitorMapTest.java

```diff
@@ -0,0 +1,36 @@
+package com.github.javaparser.utils;
+
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.visitor.ObjectIdentityEqualsVisitor;
+import com.github.javaparser.ast.visitor.ObjectIdentityHashCodeVisitor;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+
+public class VisitorMapTest {
+    @Test
+    public void normalEqualsDoesDeepCompare() {
+        CompilationUnit x1 = JavaParser.parse("class X{}");
+        CompilationUnit x2 = JavaParser.parse("class X{}");
+
+        Map<CompilationUnit, Integer> normalMap = new HashMap<>();
+        normalMap.put(x1, 1);
+        normalMap.put(x2, 2);
+        assertEquals(1, normalMap.size());
```

Comment:
@lodart - when I use a normal map, a node with the same structure is `equals` and gets the same `hashcode` so only one can be used as a map key.